server/db/interface.go

```diff
@@ -0,0 +1,186 @@
+		// Cancel order OK statuses: pending, matched, failed, executed
+		// (NOT booked, swapping, or canceled).
+		switch status {
+		case types.OrderStatusPending, types.OrderStatusMatched, types.OrderStatusFailed,
+			types.OrderStatusExecuted:
+		default:
+			return false
+		}
+
+		if ot.OrderType != order.CancelOrderType {
+			return false
+		}
+
+		// All cancel orders must have zero and filled remaining amounts.
+		if ord.Remaining() != 0 || ord.FilledAmt() != 0 {
+			panic("a CancelOrder should always return 0 Remaining and FilledAmt")
+		}
+
+	case *order.LimitOrder:
+		// Limit order OK statuses: pending, matched, swapping, booked, failed,
+		// executed, canceled (same as market plus booked).
+		switch status {
+		case types.OrderStatusPending, types.OrderStatusMatched, types.OrderStatusSwapping,
+			types.OrderStatusFailed, types.OrderStatusExecuted:
+		case types.OrderStatusBooked, types.OrderStatusCanceled:
+			// Immediate time in force limit orders may not be canceled, and may
+			// not be in the order book.
+			if ot.Force == order.ImmediateTiF {
+				return false
+			}
+		default:
+			return false
+		}
+
+		if ot.OrderType != order.LimitOrderType {
+			return false
+		}
+
+		// All limit orders must respect lot size.
+		if ot.Quantity%mkt.LotSize != 0 || ord.Remaining()%mkt.LotSize != 0 {
+			return false
+		}
+	default:
+		// cannot validate an unknown order type
+		return false
+	}
+
+	return true
+}
+
+// Orders (only validated ones):
+//  1. The essential and immutable order.Order data
+//  2. PK: Computed public order UID
+//  3. Status (canceled, failed/not matched, matched, swap in progress, swap completed)
+//  4. Remaining/filled amount (or is this a book/market item?)
+//  5. Owner account ID
+
+// Epochs:
+//  1. PK: epoch ID
+//  2. list of order IDs and order hashes (ref orders table)
+//  3. resulting shuffle order
+//  4. resulting matches (ref matches table)
+//  5. resulting order mods (change filled amount)
+//  6. resulting book mods (insert, remove)
+//  refs: orders, matches
+
+// Matches:
+//  1. The order.Match data
+//  2. Status for each maker matched with the taker (failed, executed/swapped)
+// PK is ??? a unique ID used by the Market?
+//  refs: orders
+
+// Swaps:
+//  1. The swap.Swap data
+//  2. PK: Computed public swap ID
+//  3. Status (pending, in progress, failed, executed)
+//  refs: matches
+
+// Users:
+//  1. The account.{User,Account} data
+//  2. PK: Unique user ID (server generated)
+//  3. Order IDs (or maybe just join with Orders table by UserID to get these)
+
+// Books are in-memory, but on shutdown or other maintenance events, the books can be stored.
+//  1. buy and sell orders (two differnet lists)
+
+// Market history with *timestamped* events including:
+//  1. order receipt (ref to: users table)
+//  2. order validation (ref to: orders table)
+//  3. order entry into an epoch (ref to: orders and epochs tables)
+//  4. epoch matches made (ref to: epochs table)
```

Comment:
Just reviewing that.  I see `CancelOrder(order.orderID)` is what you've specified, but I was covering that with `UpdateOrderStatus(order.Order, types.OrderStatus)` where the status is `OrderStatusCanceled`.  I'll make a `CancelOrder` function since that's far more intuitive, but an open question is if it is acceptable to to require specifying the base and quote asset to indicate which market the order pertains to.

Note that given the actual order (either concrete type or `order.Order`), the market is known.  This makes it far easier and efficient for the db backend.  That's why in the `OrderArchiver` interface I've declared functions in pairs, one taking the `order.Order` and another taking the order ID plus the base and quote asset indexes.  Ideally the DB consumer wouldn't need to provide any info beyond the order ID, but practically it is needed unless either (1) an expensive search is performed across all market's tables, or (2) a dex-wide table of order-ids is maintained that references an entry in an owning market's order table.